### ISSUE TITLE
Several characters are count twice

### ISSUE THREAD

**Issue by Tragetaschen** [dotnet/HttpRepl#472]
I'm running `5.0.2-preview.21057.1`, but I'm pretty sure this was also the case for `5.0.1`

There are several characters that count twice in terms of cursor movement.

Steps to reproduce:
- Type `C:\`
- Press `<Home>`
- Press `<End>`
- Press `<Backspace>` five times

Expected behavior: The cursor should be right before the `C` after `<Home>` and `<Backspace>` should remove three characters
Actual behavior: The cursor is within the prompt before the `>` and you can remove five characters including two from the prompt.

Steps to reproduce:
- Type `:\:\:\:\:\:\:\:\` (That is as many characters as there are in the `(Disconnected)> ` prompt)
- Press `<Backspace>` 32 times (or just hold it)

Expected behavior: The 16 characters should be removed
Actual behavior: The 16 characters as well as the 16 that form the prompt are removed and the prompt is invisible

![httprepl-characters-twice](https://user-images.githubusercontent.com/1947968/103999079-ceba5c80-519c-11eb-92c0-64926427fc65.gif)

**Comment by tlmii** [comment]
@Tragetaschen This one I can't reproduce so far. Can you give your system details? What OS (I assume Win10), what terminal (and does it act the same in, for instance, plain old cmd.exe as well as Windows Terminal and powershell), etc?

**Comment by Tragetaschen** [comment]
Interesting, I just tried it in cmd.exe, and it seems to be related to my keyboard layout.

I'm using Neo2 (https://www.neo-layout.org/) and this layout uses a couple more keys similar to Shift to reach more layers ("Ebenen"). `:` and `\` are on layer 3 and you use Caps Lock or `#` key (`\` on a US layout) to reach those characters. For `C` I type the usual `<Shift>+c`, for `:` I type `<Caps Lock>+ö` (`<Caps Lock>+:` on US).

The keyboard layout is installed as the "Windows Driver" variant, not he Auto-Hotkey (AHK) variant.

On the native cmd.exe or Windows Terminal prompt, these key combinations work as expected, and typing `C:\` works including deletion of characters. However, once I'm in `httprepl` there is a change.

Within Windows Terminal, `httprepl` shows behavior as outlined in the OP.

Within cmd.exe, I can insert what looks like a space character by pressing either `<Caps Lock>`/`#` ("Mod3") or `<AltGr>`/`<` ("Mod4"). These keys should not produce any characters, pressing those only activates higher layers in the keyboard layout.

I have just cloned the repo to gather some useful data.

**Comment by Tragetaschen** [comment]
So, pressing the "Mod3" key enters `InputManager.StartAsync` from `ReadKey` as
- Key `Oem102`
- KeyChar `\0`
- Modifier 0

Mod4 is
- Key `Oem8`
- KeyChar `\0`
- Modifier 0

The `:` character (`<Caps Lock>+ö`)
- Key `D`
- KeyChar `:`
- Modifier 0

Pressing the key labeled `ö` results in a `d` in the Neo2 layout.

**Comment by tlmii** [comment]
Ah, ok, this is making some sense. I bet the answer is somewhat related to #226 (at least partially addressed in #227). 

We have a pretty basic way of handling the keystrokes, where we make some assumptions about what produces a character (and should be passed through to the screen) and what doesn't (and thus should be checked for handlers and ignored if there's no handler). I'm betting this setup is running afoul of that somehow.

The relevant conditions start with this:

https://github.com/dotnet/HttpRepl/blob/ba8a20ccb869abf4434fc3cb09ca3ae6c3d35765/src/Microsoft.Repl/Input/InputManager.cs#L230

so that's probably where you'd want to put your breakpoints if you wanted to debug and see what your were hitting there.

**Comment by tlmii** [comment]
Without being able to repro it yet (I'll try to install neo shortly) I'd guess you can block the first two characters by adding more Unhandled registrations after this one:

https://github.com/dotnet/HttpRepl/blob/ba8a20ccb869abf4434fc3cb09ca3ae6c3d35765/src/Microsoft.Repl/Input/KeyHandlers.cs#L87

```C#
inputManager.RegistryKeyHandler(CosoleKey.Oem8, Unhandled);
inputManager.RegistryKeyHandler(CosoleKey.Oem102, Unhandled);
```

Alternatively, we could probably check for the null character here, and just ignore it:

https://github.com/dotnet/HttpRepl/blob/ba8a20ccb869abf4434fc3cb09ca3ae6c3d35765/src/Microsoft.Repl/Input/InputManager.cs#L324

```C#
if (keyPress.KeyChar == '\0') 
{
  continue;
}
```

That'd probably be more all-encompassing (and prevent us from having to mark all `Oem*` keys as `Unhandled`). But I'd worry a bit about any side effects.

**Comment by tlmii** [comment]
@Tragetaschen I _think_ this should be resolved now in master (available with the first 5.1 preview build). I didn't want to put it into 5.0.2 because it didn't quite fit as a bug fix because of the potential for breaking things if I wasn't correct on what this fix would mean to the rest of the key processing.

If you want to try building it from source, that works - otherwise it probably won't be too much longer before we get a 5.1 preview out the door. 

By way of explanation, it looks like what was happening here was our internal input buffer was including the null character (`\0`) and it was being written to the console, but the console was ignoring it. Thus the buffer and console were out of sync. So when you deleted a character from the end, visually it'd delete the right character, but internally sometimes it was deleting the `\0`, so we didn't get back to the "beginning" (where we would stop allowing deletions) on the input buffer when you'd expect based on the screen cursor position.

**Comment by Tragetaschen** [comment]
I played around with the master branch a bit and could not find anything unusual.
